### ISSUE TITLE
Reevaluate the default seccomp policy on clone and unshare

### ISSUE THREAD

**Issue by cc-wr** [moby/moby#42441]
#### Current Policy

The current policy blocks the following `CLONE_` flags for `clone` when the `CAP_SYS_ADMIN` capability is not present:

* `CLONE_NEWNS`
* `CLONE_NEWUTS`
* `CLONE_NEWIPC`
* `CLONE_NEWUSER`
* `CLONE_NEWPID`
* `CLONE_NEWNET`
* `CLONE_NEWCGROUP`

https://github.com/moby/moby/blob/a8a769f04f7ae44117772943ef0f4607b54e05ff/profiles/seccomp/default_linux.go#L563

Without this policy, the kernel would have blocked all of the above flags *except* `CLONE_NEWUSER` (since Linux 3.8) if the `CAP_SYS_ADMIN` capability was not present.

---

The current policy also completely blocks `unshare` when the `CAP_SYS_ADMIN` capability is not present:

https://github.com/moby/moby/blob/a8a769f04f7ae44117772943ef0f4607b54e05ff/profiles/seccomp/default_linux.go#L547

Without this policy, the kernel would have _allowed_ the following `CLONE_` flags for `unshare` even if the `CAP_SYS_ADMIN` capability was not present:

* `CLONE_FILES`
* `CLONE_FS`
* `CLONE_SYSVSEM`
* `CLONE_NEWUSER`

#### Some Proposed Changes

As the unprivileged creation of user namespaces becomes more secure, I am wondering if Docker should continue to block it by default. It may be sensible to no longer disallow the `CLONE_NEWUSER` flag when the `CAP_SYS_ADMIN` capability is not present. If it made sense, we could also only allow unprivileged `CLONE_NEWUSER` for newer Linux versions using seccomp's `minKernel`.

I believe this change would also make it easier for users to mount FUSE filesystems without the `CAP_SYS_ADMIN` capability. For example, this change should allow the following workflow, as long as `--device /dev/fuse` is also provided:

1. Create user namespace
2. With `CAP_SYS_ADMIN` capability (in the user namespace), ...
        3. Create mount namespace
        4. Mount FUSE filesystem (allowed since Linux 4.18)

Some related events:

* [Debian now _allows_ the unprivileged creation of user namespaces by default](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=898446)
* [RHEL seems to now _allow_ the unprivileged creation of user namespaces by default](https://www.stigviewer.com/stig/red_hat_enterprise_linux_8/2020-11-25/finding/V-230548)
* [CRI-O's default seccomp profile seems to _allow_ `clone` and `unshare` without any restriction](https://github.com/containers/common/blob/master/pkg/seccomp/seccomp.json)

---

I also believe the default policy with respect to `unshare` could become more fine-tuned. As mentioned above, there are several flags for `unshare` that the kernel allows without `CAP_SYS_ADMIN` that we currently block anyway. However, with these specific flags (except `CLONE_NEWUSER`), users could just call `clone` *without* a given flag to achieve the same result as calling `unshare` *with* that flag, and it would be allowed, so the current restriction on `unshare` seems like it might be able to be improved.

My proposal would be that we block `unshare` when the `CAP_SYS_ADMIN` capability is not present, except for the first three flags I listed above (and `CLONE_NEWUSER` if we allow it for `clone`).

#### Some Related Discussions

* https://github.com/moby/moby/pull/41880
* https://github.com/docker/for-linux/issues/321
* https://github.com/moby/moby/pull/41244
* https://github.com/containers/buildah/issues/1901

**Comment by AkihiroSuda** [comment]
Can we have this as a non-default built-in profile like `--security-opt seccomp=allow-unshare-user`?

Or if we are going to have this as the default, we will need to provide `seccomp=disallow-unshare-user` option.

**Comment by cc-wr** [comment]
My vote would be for allowing `CLONE_NEWUSER` as the default. As user namespaces increasingly become the standard way of achieving certain things as an unprivileged user — like mounting FUSE filesystems, or more effective chrooting (requires the currently blocked `pivot_root`, but you get the point) — it may be overly restrictive to continue to disallow it. I think there is similar discussion in the Debian ticket I linked.

**Comment by cc-wr** [comment]
> However, with these specific flags (except `CLONE_NEWUSER`), users could just call clone without a given flag to achieve the same result as calling `unshare` with that flag, and it would be allowed, so the current restriction on `unshare` seems like it might be able to be improved.

As a correction, I believe this part of my description was wrong. I did not realize unprivileged users were blocked from directly calling `clone` entirely.

**Comment by rossburton** [comment]
Any progress on this?  I really want to `unshare -run` inside a docker without having to use a privileged container.

**Comment by adamnovak** [comment]
I believe that changing the default here would make `singularity` able to run nested containers inside unprivileged Docker containers on default deployments, using its `--userns` option.

**Comment by skissane** [comment]
The discussion in this issue has focused primarily on `CLONE_NEWUSER`. However, there is another issue here, which is `CLONE_FS`. In Unix(-like) systems, the current directory is defined as per-process; for various reasons, some people want a per-thread current directory. On macOS, there is an (undocumented) API to do that: `pthread_chdir_np`/`pthread_fchdir_np`. That doesn't exist on Linux, but the equivalent is to do `unshare(CLONE_FS)`, which disconnects the filesystem attributes (current directory, root directory and umask) of the calling thread from the rest of the process, and a subsequent call to `chdir`/`fchdir` will only affect that thread. However, that is prohibited by Docker's default seccomp profile, unless you have `CAP_SYS_ADMIN`.

I understand part of the reason for that, was to block `CLONE_NEWUSER ` due to concern about potential security bugs related to user namespaces. But, I'm not aware of any equivalent concerns regarding un-sharing `CLONE_FS`, which has significantly less security implications – it just lets a thread have a separate working directory and umask from the other threads in the process. (It also lets it have its own root directory, which is more security-sensitive–but `unshare(CLONE_FS)` doesn't change the privileges needed to call `chroot`, so the potential security implications seem rather limited.)

Given that, if there is any security-related hesitancy about unblocking `CLONE_NEWUSER`, maybe `CLONE_FS` could be unblocked sooner? Should I create a separate issue for that specifically?

One reason why one might want a per-thread current directory, is to work around the 108 character path name limitation of Unix domain sockets. If you use `chdir` with a relative path, then the 108 character limit only applies to the file name, not the full path. However, in a multithreaded process, `chdir` is not thread safe, since it impacts all threads in the process; using a per-thread current directory, one can use a separate thread to do the socket operation, and thus avoid any race condition bugs.  FreeBSD has a much nicer solution to this – `bindat`/`connectat` system calls, which take a directory file descriptor for resolving relative paths, just like what `openat`/etc do – however, thus far nobody has ported that to Linux, so `unshare(CLONE_FS)` is an alternative.

Note a workaround to this blocking is to just skip `pthread_create` and manually call `clone` without the `CLONE_FS` flag. However, manually calling `clone` is more work and complexity than just using `pthread_create`. And it doesn't make a lot of sense to have a seccomp profile that lets us unshare `CLONE_FS` at thread creation time (by omitting it in the `clone` call), but not after the thread is started.

There is a similar argument that blocking `unshare(CLONE_FILES)` and `unshare(CLONE_SYSVSEM)` is a bit silly given it only stops you from unsharing them after the thread is started, not at thread creation time (by omitting them from the clone flags). However, unlike `unshare(CLONE_FS)`, I don't personally have a use case for either of those (but maybe someone else does.)

**Comment by slonopotamus** [comment]
One more argument for allowing `unshare(CLONE_NEWUSER)`: [buildah](https://github.com/containers/buildah) depends on it and simply doesn't work under default Docker seccomp policy.

**Comment by slonopotamus** [comment]
See https://github.com/moby/profiles/pull/4

**Comment by sbrivio-rh** [comment]
...and https://github.com/moby/profiles/pull/5, even though it's a partial duplicate of https://github.com/moby/profiles/pull/4.
